### PR TITLE
Feature/optimise step selector builder

### PR DESCRIPTION
**Pull request by rutefig**
### Description
- **Goal:** is to optimise the Step Selector Builder to have Log2(n) Columns per n Step Types instead of one Column per each Step Type.
- **Current Implementation:** implemented a new type of Selector and implemented the build function for it

- **Explanation:** If we use binary representation for the step types instead of using the actual columns we will get an optimised version for the Selector Builder since we can get the total of columns from the following expression
```math
n\_cols = \lceil \log_2(n\_step\_types + 1) \rceil
```
We need take into consideration that we also need to save one binary value for the case when there is no step type, because of this we do the math with \(n\_step\_types + 1\).

 ### What is missing
- [x] Ensure compatibility with the backend
- [x] Unit tests

### CODE REVIEW

**Comment by leolara** [comment]
4 steps require 3 columns, because 00 is not used. Please add a test with 4 steps that checks that 3 columns are used.

File: src/plonkish/compiler/step_selector.rs [hunk 1]
```diff
@@ -197,3 +259,137 @@ fn other_step_type<F>(unit: &CompilationUnit<F>, uuid: UUID) -> Option<Rc<StepTy
...
+                    Uuid::now_v1(&[1, 2, 3, 4, 5, 6]).as_u128(),
+                    format!("StepType{}", i),
+                )),
+            );
+        }
+
+        builder.build(&mut unit);
+
+        // Assert number of columns
+        let n_cols = (n_step_types as f64).log2().ceil() as usize;
```
Comment (leolara) [inline src/plonkish/compiler/step_selector.rs:386]: +1 missing

File: src/plonkish/compiler/step_selector.rs [hunk 2]
```diff
@@ -197,3 +259,137 @@ fn other_step_type<F>(unit: &CompilationUnit<F>, uuid: UUID) -> Option<Rc<StepTy
...
+            )),
+        );
+
+        builder.build(&mut unit);
+
+        // Assert number of columns
+        assert_eq!(unit.columns.len(), 3);
+        assert_eq!(unit.selector.columns.len(), 3);
+
+        // TODO: Assert expressions
```
Comment (leolara) [inline src/plonkish/compiler/step_selector.rs:358]: @rutefig do TODOs

**Comment by leolara** [comment]
Rusfmt problems

**Comment by leolara** [comment]
I think you test the expresions and the assignments, the expressions can be converted to a string with `print!(":#?", expr) ` and then compared to a string @rutefig

File: src/plonkish/compiler/step_selector.rs [hunk 3]
```diff
@@ -188,6 +188,69 @@ impl StepSelectorBuilder for TwoStepsSelectorBuilder {
...
+                    combined_expr_not = combined_expr_not
+                        * (PolyExpr::Const(F::ONE) - column.query(0, format!("Column {}", i)));
+                    assignments.push((column.query(0, format!("Column {}", i)), F::ONE));
+                } else {
+                    combined_expr = combined_expr
+                        * (PolyExpr::Const(F::ONE) - column.query(0, format!("Column {}", i)));
+                    combined_expr_not =
+                        combined_expr_not * column.query(0, format!("Column {}", i));
+                }
+            }
```
Comment (leolara) [inline src/plonkish/compiler/step_selector.rs:235]: @rutefig I think the right combined_expr_not is always (1-combined_expr)

**Review by leolara — APPROVED** [review]

**Comment by rutefig** [comment]
@leolara need approval here, the previous run failed on the Clippy so I committed a fix for that

**Review by leolara — APPROVED** [review]

**Pull request merged**